activesupport/lib/active_support/actionable_error.rb

```diff
@@ -43,6 +55,23 @@ module ClassMethods
       def action(name, &block)
         _actions[name] = block
       end
+
+      # Trigger the current actionable error from a pre-existing if the
+      # condition, given as a block, returns a truthy value.
+      #
+      #   class SetupError < Error
+      #     include ActiveSupport::ActionableError
+      #
+      #     trigger on: ActiveRecord::RecordInvalid, if: -> error do
+      #       error.to_s.match?(InboundEmail.table_name)
+      #     end
+      #   end
+      def trigger(options)
+        error = options.fetch(:on) { raise ArgumentError, "missing keyword: on" }
```

Comment:
Would it still not be more helpful to explicitly check that `error` is an `Exception` at invocation time, rather than allowing the action to never be triggered due to a bad constant that was passed for `on:`?

Reply:
I'm not sure how defensive we want to be here. The `trigger` method is used in the context of a class like:

```ruby
class InstallError
  trigger on: StandardError, if: condition
  # You probably won't be rescuing an error in the body of a class.
end
```

In the body of the class, you probably won't be catching already risen errors, but you'd have access to the exception classes themselves. Guarding for this case would make the code more verbose, while I don't currently see this as a common mistake to need the guard and make the check's worth, in a way. I can change the naming, though, to prevent any confusion here.

Reply:
Sorry, I am not sure I follow your response and I suspect we could be talking about different things. Let me try to clarify what I mean. 

If I were to do something like this:
```ruby
class StupidError < StandardError
  include ActiveSupport::ActionableError

  trigger on: ActiveModel, if: condition
end
```
then this will never work since I did not pass a subclass of `Exception` or `StandardError` to the `on` parameter of the `trigger` method.

However, nothing will give me any indication that this will never work either as the PR currently stands. I am just suggesting that `trigger` validates that what is passed to `on:` is a subclass of `StandardError` and raises an error if it is not.

Reply:
If you want to break it, sure, you can! However, I'd prefer if we're not so defensive here.

Ruby's waters are quite muddy when it comes to what can rescue an exception. For example, the following compiles just fine:

```ruby
begin
  raise "Runtime Error"
rescue String
  p "Hey, I'm never called!"
end

#=> nil
```

Of course, it does nothing, but proves a point. 😅 Ruby's rescue doesn't match errors strictly based on their type hierarchy. It matches the errors with the case operator (`===`) and this let's you do interesting things like:

```ruby
def String.===(other)
  RuntimeError === other or super
end

begin
  raise "Runtime Error"
rescue String
  p "Look, Ma, I rescue String's (RuntimeError) now!"
end

#=> "Look, Ma, I rescue String's (RuntimeError) now!"
```

The only check `rescue` does to its rhs is that it's a module, not an `Exception` descendant per se. If its a module, it uses the `===` operator and you can have dynamic error catchers. They are surprisingly useful if you want to suppress noisy errors that don't adhere to a specific hierarchy, but you can still bring them together under one umbrella, or... it let's you rescue errors even based on their `#message` or other properties, if you inherit a messy codebase. _(See [masse](https://github.com/gsamokovarov/masse) for example.)_

If we want to be really correct in guarding the `trigger on:`  case, we'll have to write all those checks and I don't see any clear win for us here. If it was something that's called often and would have blown up mystically later on, maybe I would have gone the extra mile and guard against such cases. However, the way I see the actionable errors trigger usage at the moment is as follows.

You define it once, you test that the error is triggered and... everything is kinda static after that. The triggers (and the actionable errors) are not dynamically defined and probably won't be used that much. So, if you get the error to trigger once, you're good. You have to test the actual trigger code more, not just the error you passed to it.

I also want to keep the door open for dynamic error matcher support* and  the following interface:

```ruby
trigger on: 'ActiveRecord::StatementInvalid", if: -> error do
  # ...
end
```

This could be useful if we want to delay the autoloading of `ActiveRecord::StatementInvalid` should the current actionable errors load Active Record too soon.

_* Currently, the triggers doesn't match the errors with `#===` for simplicitly, but if someone needs it, we can support it later on._

Reply:
Ok, convinced! 😄 Thank you for that extensive and well-written explanation, makes it crystal clear why such a check might be unnecessary! 👍